pallets/proxy-teleport/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+			message_id: XcmHash,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(10_000)]
+		pub fn proxy_native_teleport(
+			origin: OriginFor<T>,
+			dest: Box<VersionedMultiLocation>,
+			beneficiary: Box<VersionedMultiLocation>,
+			native_asset_amount: u128,
+			proxy_asset: Box<VersionedMultiAssets>,
+			fee_asset_item: u32,
+		) -> DispatchResult {
+			Self::do_proxy_teleport_assets(
+				origin,
+				dest,
+				beneficiary,
+				native_asset_amount,
+				proxy_asset,
+				fee_asset_item,
+			)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn do_proxy_teleport_assets(
+		origin: OriginFor<T>,
+		dest: Box<VersionedMultiLocation>,
+		beneficiary: Box<VersionedMultiLocation>,
+		native_asset_amount: u128,
+		proxy_asset: Box<VersionedMultiAssets>,
+		fee_asset_item: u32,
+	) -> DispatchResult {
+		//Unbox origin, destination and beneficiary.
+		let origin_location = T::ExecuteXcmOrigin::ensure_origin(origin)?;
+		let dest: MultiLocation =
+			(*dest).try_into().map_err(|()| pallet_xcm::Error::<T>::BadVersion)?;
+		let beneficiary: MultiLocation =
+			(*beneficiary).try_into().map_err(|()| pallet_xcm::Error::<T>::BadVersion)?;
+
+		//Create assets
+
+		// Native from local perspective
+		let native_asset = MultiAsset {
+			id: AssetId::Concrete(MultiLocation::here()),
+			fun: Fungibility::Fungible(native_asset_amount),
+		};
+
+		let assets = MultiAssets::from(vec![native_asset]);
+
+		// Native from foreign perspective
+		//TODO: Replace ID with parameter
+		let localtion_as_foreign =
+			MultiLocation { parents: 1, interior: Junctions::X1(Junction::Parachain(1836)) };
+
+		let native_as_foreign = MultiAsset {
+			id: AssetId::Concrete(localtion_as_foreign),
+			fun: Fungibility::Fungible(native_asset_amount),
+		};
+
+		let foreing_assets = MultiAssets::from(vec![native_as_foreign]);
+
+		//Unbox proxy asset
+		let proxy_asset: MultiAssets =
+			(*proxy_asset).try_into().map_err(|()| pallet_xcm::Error::<T>::BadVersion)?;
+
+		//TeleportFilter check
+		let value = (origin_location, assets.into_inner());
+		ensure!(T::XcmTeleportFilter::contains(&value), pallet_xcm::Error::<T>::Filtered);
+		let (origin_location, assets) = value;
+
+		// Reanchor the proxy asset to the destination chain.
+		let context = T::UniversalLocation::get();
+		let fees = proxy_asset
+			.get(fee_asset_item as usize)
+			.ok_or(pallet_xcm::Error::<T>::Empty)?
+			.clone()
+			.reanchored(&dest, context)
+			.map_err(|_| pallet_xcm::Error::<T>::CannotReanchor)?;
+
+		// TODO: Define if Withdrawn proxy assets are deposited or trapped. 
+		// Check if there is no vulnerability through RefundSurplus
+		//let max_assets = (assets.len() as u32).checked_add(1).ok_or(Error::<T>::TooManyAssets)?;
+
+		//Build the message to execute on origin.
+		let assets: MultiAssets = assets.into();
+		let message: Xcm<<T as frame_system::Config>::RuntimeCall> = Xcm(vec![
+			// Withdraw drops asset so is used as burn mechanism
```

Comment:
I was wondering, is it possible to mimic [what `InitiateTeleport` does](https://github.com/paritytech/polkadot/blob/962bc21352f5f80a580db5a28d05154ede4a9f86/xcm/xcm-executor/src/lib.rs#L646) so that we don't leave the native asset trapped in the holding?

Reply:
This way we could directly pass `assets` in line 172 instead of the hardcoded `foreing_assets`.

Reply:
I like this, on it! Thanks

Reply:
Update:
- I have implemented the [`reanchor` ](https://github.com/paritytech/polkadot/blob/962bc21352f5f80a580db5a28d05154ede4a9f86/xcm/xcm-executor/src/lib.rs#L661C4-L661C5), `foreign_assets` is no longer hardcoded.
- I have also added a `BurnAssets` to mimic [this part](https://github.com/paritytech/polkadot/blob/962bc21352f5f80a580db5a28d05154ede4a9f86/xcm/xcm-executor/src/lib.rs#L648C3-L648C3) of the `InitiateTeleport` instruction 
- Currently trying to solve [this `check_out` logic](https://github.com/paritytech/polkadot/blob/962bc21352f5f80a580db5a28d05154ede4a9f86/xcm/xcm-executor/src/lib.rs#L657C20-L657C20) for this proxy call to track teleports too.

Reply:
Good idea to use `BurnAsssets` instruction. In my opinion this is a solution as clean as we can get. This way there is no trapped assets in the holding whereas the counterpart is expected to be minted in the destination chain. I think the check out logic is then secondary.